### ISSUE TITLE
Adding More Models to Modelzoo

### ISSUE THREAD

**Issue by colah** [tensorflow/lucid#86]
I'd really like get more models into lucid modelzoo.

# Background

It seems like there are two major reservoirs of pre-trained models: [tf slim](https://github.com/tensorflow/models/tree/master/research/slim#pre-trained-models) and [caffe](https://github.com/BVLC/caffe/wiki/Model-Zoo). Most pre-trained models online come from one of those. In many cases, there are two different pre-trained models going by the same name floating around, because both projects re-implemented the model. (tf-hub seems to mostly be repackaging a subset of the tf-slim models.)

Typically, neither of these sources is the canonical model.

# Questions / Considerations

**Should we really be creating another model zoo, or is there a way we could just pull from an existing one?**

* Only tf-hub seems provide models in an easy to import way; it's mostly a subset of slim.
* We want models from different sources
* We may want more meta-data than others

**How do we deal with multiple different implementations of the "same" model?**

* Post-fix names (eg. `InceptionV3_slim` ?)

**It would be really nice to have the same model trained on multiple datasets for some experiments**

* https://github.com/CSAILVision/places365

**What meta-data should we be including about models?**

**Comment by colah** [comment]
Initial work on importing models (eg. AlexNet and InceptionV3) in #85.

**Comment by colah** [comment]
# Slim -> Frozen

Example for Inception V3 (from [slim documentation](https://github.com/tensorflow/models/tree/master/research/slim#pre-trained-models))

```sh
git clone https://github.com/tensorflow/models.git
wget https://raw.githubusercontent.com/tensorflow/tensorflow/master/tensorflow/python/tools/freeze_graph.py

python models/research/slim/export_inference_graph.py \
  --alsologtostderr \
  --model_name=inception_v3 \
  --output_file=/tmp/inception_v3_inf_graph.pb

CHECKPOINT_DIR=/tmp/checkpoints
mkdir ${CHECKPOINT_DIR}
wget http://download.tensorflow.org/models/inception_v3_2016_08_28.tar.gz
tar -xvf inception_v3_2016_08_28.tar.gz
mv inception_v3.ckpt ${CHECKPOINT_DIR}
rm inception_v3_2016_08_28.tar.gz

python freeze_graph.py \
  --input_graph=/tmp/inception_v3_inf_graph.pb \
  --input_checkpoint=/tmp/checkpoints/inception_v3.ckpt \
  --input_binary=true --output_graph=/tmp/frozen_inception_v3.pb \
  --output_node_names=InceptionV3/Predictions/Reshape_1
```

# Caffe -> Frozen

Microsoft's new [MMdnn](https://github.com/Microsoft/MMdnn) tool is a promising way to convert models to TensorFlow. There was an issue (Microsoft/MMdnn/#117) where someone was trying to do Caffe -> Frozen like we want to.

**Comment by colah** [comment]
# Modelzoo Resources

Eventually, it would be nice to have a list of models with:

* model graphs
* pre-computed neuron visualizations
* links to papers

**Comment by colah** [comment]
I've been trying to mass export the tensorflow slim models. It appears that some of the provided checkpoints are not compatible with their present code (tensorflow/models#4938).

On the plus side, I have successfully exported 8 more models! :)

**Comment by colah** [comment]
Also, I've got several mostrostities of an ipython-bash fusion script to automatically export models from tf-slim to us. Uh, is that some we'd want to... check in somewhere?? @ludwigschubert ?

More generally, the workflows for getting models from different formats into something sane is full of gotchas and workarounds. For now, I'm using this issue as a bit of a log, but it would be good to have institutional memory about how to do this.

**Comment by ludwigschubert** [comment]
My writeup was in [this notebook](https://github.com/tensorflow/lucid/issues/34#issuecomment-376663839), primarily because I wanted it to be easy for users to put in their own graph and still follow the steps. Let's merge both of our attempts when you have reached a plateau in adding new models.

**Comment by colah** [comment]
## Update: merged #85, adding more models

Despite there likely being some bugs slipping through, I've merged #85. It adds many more models, both from tf-slim and caffe (see list below).

Some potential problems -- at the moment, I'm inferring all of these from the convention of the framework the model was trained in, but this may not be reliable:

* ⚠️ Some models may not have the correct list of labels.
* ⚠️ Presently assuming all caffe models are BGR but it's possible some aren't in cases where they were originally trained in a different framework.
* ⚠️ Some models may not have the correct input range.
* 🏗️ Layer lists for some models are empty.

There are more caffe models that we did not merge.

In addition to the models, the new PR adds scripts at [lucid/scratch/scripts/import_slim](https://github.com/tensorflow/lucid/tree/master/lucid/scratch/scripts/import_slim) and [lucid/scratch/scripts/import_caffe](https://github.com/tensorflow/lucid/tree/master/lucid/scratch/scripts/import_caffe) demonstrating how to import models of those types.

### TF Slim Models

* ✅  **inception_v1**
* ✅  **inception_v2**
* ✅  **inception_v3**
* ✅  **inception_v4**
* ✅  **inception_resnet_v2**
* ✅  **resnet_v1_50**
* ✅  **resnet_v1_101**
* ✅  **resnet_v1_152**
* ✅  **resnet_v2_50**
* ✅  **resnet_v2_101**
* ✅  **resnet_v2_152**
* ✅  **vgg_16**
* ✅  **vgg_19**
* ✅  **mobilenet_v1**
* ✅  **mobilenet_v1_050**
* ✅  **mobilenet_v1_025**
* ✅  **nasnet-a_mobile**
* ✅  **nasnet-a_large**
* ✅  **pnasnet-5_large**
* ✅  **pnasnet-5_mobile**

### Caffe Models

* ✅  **AlexNet**
* ✅ [**AlexNet (Places 365)**](https://github.com/CSAILVision/places365)
* ✅  **InceptionV1**
* ✅  **InceptionV1 (Places 205)**
* ✅  **InceptionV1 (Places 365)**
* ✅  [**CaffeNet**](https://github.com/BVLC/caffe/tree/master/models/bvlc_reference_caffenet)
* ✅  **VGG_16**
* ✅  **VGG_19**
